### PR TITLE
Move more functions to db.Find

### PR DESCRIPTION
**Pull request by lunny**
Following #28220

This PR move more functions to use `db.Find`.

### CODE REVIEW

**Comment by delvh** [comment]
Something seems to have changed:
> 2023/12/13 09:36:03 ...hook_verification.go:68:readAndVerifyCommitsFromShaReader() [E] Unverified commit: d766f2917716d45be24bfa968b8409544941be32
2023/12/13 09:36:03 ...hook_verification.go:49:func2() [E] Unverified commit: d766f2917716d45be24bfa968b8409544941be32
2023/12/13 09:36:03 ...hook_verification.go:68:readAndVerifyCommitsFromShaReader() [E] Unverified commit: 93eac826f6188f34646cea81bf426aa5ba7d3bfe
2023/12/13 09:36:03 ...hook_verification.go:49:func2() [E] Unverified commit: 93eac826f6188f34646cea81bf426aa5ba7d3bfe
2023/12/13 09:36:03 ...hook_verification.go:68:readAndVerifyCommitsFromShaReader() [E] Unverified commit: 72920278f2f999e3005801e5d5b8ab8139d3641c
2023/12/13 09:36:03 ...hook_verification.go:49:func2() [E] Unverified commit: 72920278f2f999e3005801e5d5b8ab8139d3641c
2023/12/13 09:36:03 ...hook_verification.go:68:readAndVerifyCommitsFromShaReader() [E] Unverified commit: 9ce3f779ae33f31fce17fac3c512047b75d7498b
2023/12/13 09:36:03 ...hook_verification.go:49:func2() [E] Unverified commit: 9ce3f779ae33f31fce17fac3c512047b75d7498b
--- FAIL: TestVerifyCommits (0.18s)
    hook_verification_test.go:38: 
        	Error Trace:	/home/runner/work/gitea/gitea/routers/private/hook_verification_test.go:38
        	Error:      	Received unexpected error:
        	            	Unverified commit: d766f2917716d45be24bfa968b8409544941be32
        	Test:       	TestVerifyCommits
    hook_verification_test.go:38: 
        	Error Trace:	/home/runner/work/gitea/gitea/routers/private/hook_verification_test.go:38
        	Error:      	Received unexpected error:
        	            	Unverified commit: 93eac826f6188f34646cea81bf426aa5ba7d3bfe
        	Test:       	TestVerifyCommits

**Comment by delvh** [comment]
Sounds a bit like this PR is doomed to never be merged?
After all, the CI seems to have unfixable problems…

**Comment by lunny** [comment]
Maybe blocked by #28659

File: routers/api/v1/repo/pull.go [hunk 1]
```diff
@@ -1329,17 +1329,20 @@ func GetPullRequestCommits(ctx *context.APIContext) {
...
+	start, limit := listOptions.GetSkipTake()
 
-	if end > totalNumberOfCommits {
-		end = totalNumberOfCommits
+	if start+limit > totalNumberOfCommits {
+		limit = totalNumberOfCommits - start
+	}
+	if limit < 0 {
+		limit = 0
 	}
```
Comment (delvh) [inline routers/api/v1/repo/pull.go]:
```suggestion
	limit = max(limit, 0)
```

File: routers/api/v1/repo/pull.go [hunk 2]
```diff
@@ -1329,17 +1329,20 @@ func GetPullRequestCommits(ctx *context.APIContext) {
...
 	userCache := make(map[string]*user_model.User)
 
-	start, end := listOptions.GetStartEnd()
+	start, limit := listOptions.GetSkipTake()
 
-	if end > totalNumberOfCommits {
-		end = totalNumberOfCommits
+	if start+limit > totalNumberOfCommits {
+		limit = totalNumberOfCommits - start
+	}
```
Comment (delvh) [inline routers/api/v1/repo/pull.go]:
```suggestion
	limit = min(limit, totalNumberOfCommits-start)
```

File: routers/api/v1/repo/pull.go [hunk 3]
```diff
@@ -1477,19 +1480,18 @@ func GetPullRequestFiles(ctx *context.APIContext) {
...
 	totalNumberOfPages := int(math.Ceil(float64(totalNumberOfFiles) / float64(listOptions.PageSize)))
 
-	start, end := listOptions.GetStartEnd()
+	start, limit := listOptions.GetSkipTake()
 
-	if end > totalNumberOfFiles {
-		end = totalNumberOfFiles
+	if start+limit > totalNumberOfFiles {
+		limit = totalNumberOfFiles - start
 	}
```
Comment (delvh) [inline routers/api/v1/repo/pull.go]:
```suggestion
	limit = min(limit, totalNumberOfFiles - start)
```

File: routers/api/v1/repo/pull.go [hunk 4]
```diff
@@ -1477,19 +1480,18 @@ func GetPullRequestFiles(ctx *context.APIContext) {
...
 
-	lenFiles := end - start
-	if lenFiles < 0 {
-		lenFiles = 0
+	if limit < 0 {
+		limit = 0
 	}
```
Comment (delvh) [inline routers/api/v1/repo/pull.go]:
```suggestion
	limit = max(limit, 0)
```

File: models/asymkey/gpg_key_list.go [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+	ids := make([]string, len(keys))
+	for i, key := range keys {
+		ids[i] = key.KeyID
+	}
+	return ids
+}
+
+func (keys GPGKeyList) LoadSubKeys(ctx context.Context) error {
+	subKeys := make([]*GPGKey, 0, len(keys))
+	if err := db.GetEngine(ctx).In("primary_key_id", keys.keyIDs()).Find(&subKeys); err != nil {
```
Comment (delvh) [inline models/asymkey/gpg_key_list.go:24]: `PrimaryKeyID` should be an `INDEX`, which it currently isn't.

Comment (lunny) [reply models/asymkey/gpg_key_list.go:24]: Yes, but it should be another PR.

**Review by delvh — APPROVED** [review]
`make fmt`

**Comment by delvh** [comment]
Ping for other reviewers

**Comment by lunny** [comment]
Last call @go-gitea/technical-oversight-committee

File: models/asymkey/ssh_key.go [hunk 6]
```diff
@@ -197,7 +197,11 @@ func (opts FindPublicKeyOptions) ToConds() builder.Cond {
 		cond = cond.And(builder.Eq{"fingerprint": opts.Fingerprint})
 	}
 	if len(opts.KeyTypes) > 0 {
-		cond = cond.And(builder.In("type", opts.KeyTypes))
+		if len(opts.KeyTypes) == 1 {
+			cond = cond.And(builder.Eq{"type": opts.KeyTypes[0]})
+		} else {
+			cond = cond.And(builder.In("type", opts.KeyTypes))
+		}
```
Comment (KN4CK3R) [inline models/asymkey/ssh_key.go]: Do we want such micro-optimization in the code? I would expect that the database can optimize such cases. MSSQL for example generates the same query plan for both.

File: models/issues/tracked_time.go [hunk 7]
```diff
@@ -117,17 +117,29 @@ func (opts *FindTrackedTimesOptions) toCond() builder.Cond {
 	return cond
 }
 
+func (opts FindTrackedTimesOptions) ToJoins() []db.JoinFunc {
```
Comment (KN4CK3R) [inline models/issues/tracked_time.go]:
```suggestion
func (opts *FindTrackedTimesOptions) ToJoins() []db.JoinFunc {
```

File: services/repository/push.go [hunk 8]
```diff
@@ -327,7 +327,10 @@ func pushUpdateAddTags(ctx context.Context, repo *repo_model.Repository, gitRepo
 		lowerTags = append(lowerTags, strings.ToLower(tag))
 	}
 
-	releases, err := repo_model.GetReleasesByRepoIDAndNames(ctx, repo.ID, lowerTags)
+	releases, err := db.Find[repo_model.Release](ctx, repo_model.FindReleasesOptions{
+		RepoID:   repo.ID,
+		TagNames: lowerTags,
+	})
 	if err != nil {
 		return fmt.Errorf("GetReleasesByRepoIDAndNames: %w", err)
```
Comment (KN4CK3R) [inline services/repository/push.go]: What about the error messages?

File: services/asymkey/sign.go
Comment (KN4CK3R) [inline services/asymkey/sign.go:1]: I'm not sure if this is really a step in the right direction. Now a caller needs to know that `IsPrimaryKeyIDEmpty` must be `true` to get the correct result. The helper in the models hides such implementation details.

File: models/repo/collaboration.go [hunk 9]
```diff
@@ -37,35 +38,34 @@ type Collaborator struct {
...
-			if user_model.IsErrUserNotExist(err) {
-				log.Warn("Inconsistent DB: User: %d is listed as collaborator of %-v but does not exist", c.UserID, repoID)
-				user = user_model.NewGhostUser()
-			} else {
-				return nil, err
-			}
-		}
 		collaborators = append(collaborators, &Collaborator{
-			User:          user,
+			User:          usersMap[c.UserID],
```
Comment (KN4CK3R) [inline models/repo/collaboration.go]: The old code used the Ghost user if the user was not found. Now `User` is nil.

File: routers/web/user/home.go [hunk 10]
```diff
@@ -663,12 +663,21 @@ func ShowSSHKeys(ctx *context.Context) {
...
+	})
 	if err != nil {
 		ctx.ServerError("ListGPGKeys", err)
 		return
 	}
 
+	if err := asymkey_model.GPGKeyList(keys).LoadSubKeys(ctx); err != nil {
+		ctx.ServerError("LoadSubKeys", err)
+		return
+	}
```
Comment (KN4CK3R) [inline routers/web/user/home.go]: Where are the sub keys used here?

File: models/asymkey/ssh_key.go (hunk 6 above)
Comment (lunny) [reply models/asymkey/ssh_key.go]: I cannot expect every database Gitea supported can do the optimization themselves.

File: services/asymkey/sign.go
Comment (delvh) [reply services/asymkey/sign.go:1]: I agree, whenever there is additional logic, we should keep the extra function.

File: models/asymkey/ssh_key.go (hunk 6 above)
Comment (KN4CK3R) [reply models/asymkey/ssh_key.go]: I would agree if this code would be in the hot path of execution. We don't do this anywhere else in the code. I have tested SQLite, MSSQL and Postgres, they all generate the same execution plan for both queries. I expect MySQL and Oracle to do the same but I don't have access to these at the moment. More databases are not supported by Gitea.

File: services/asymkey/sign.go
Comment (lunny) [reply services/asymkey/sign.go:1]: Now I changed the options logic. Please review again.

File: models/asymkey/ssh_key.go (hunk 6 above)
Comment (lunny) [reply models/asymkey/ssh_key.go]: reverted.

File: routers/api/v1/user/gpg_key.go [hunk 11]
```diff
@@ -18,23 +18,26 @@ import (
 )
 
 func listGPGKeys(ctx *context.APIContext, uid int64, listOptions db.ListOptions) {
-	keys, err := asymkey_model.ListGPGKeys(ctx, uid, listOptions)
+	opts := asymkey_model.FindGPGKeyOptions{
+		ListOptions: listOptions,
+		OwnerID:     uid,
+	}
+	keys, total, err := db.FindAndCount[asymkey_model.GPGKey](ctx, opts)
```
Comment (denyskon) [inline routers/api/v1/user/gpg_key.go]:
```suggestion
	keys, total, err := db.FindAndCount[asymkey_model.GPGKey](ctx, asymkey_model.FindGPGKeyOptions{
		ListOptions: listOptions,
		OwnerID:     uid,
	})
```

File: tests/integration/mirror_pull_test.go [hunk 12]
```diff
@@ -58,8 +58,12 @@ func TestMirrorPull(t *testing.T) {
...
 	defer gitRepo.Close()
 
-	findOptions := repo_model.FindReleasesOptions{IncludeDrafts: true, IncludeTags: true}
-	initCount, err := repo_model.GetReleaseCountByRepoID(db.DefaultContext, mirror.ID, findOptions)
+	findOptions := repo_model.FindReleasesOptions{
+		IncludeDrafts: true,
+		IncludeTags:   true,
+		RepoID:        mirror.ID,
+	}
+	initCount, err := db.Count[repo_model.Release](db.DefaultContext, findOptions)
```
Comment (denyskon) [inline tests/integration/mirror_pull_test.go:66]:
```suggestion
	initCount, err := db.Count[repo_model.Release](db.DefaultContext, repo_model.FindReleasesOptions{
		IncludeDrafts: true,
		IncludeTags:   true,
		RepoID:        mirror.ID,
	})
```

File: models/repo/collaboration.go [hunk 13]
```diff
@@ -37,35 +38,38 @@ type Collaborator struct {
...
-				user = user_model.NewGhostUser()
-			} else {
-				return nil, err
-			}
+		userIDs = append(userIDs, c.UserID)
+	}
+
+	usersMap := make(map[int64]*user_model.User)
+	if err := db.GetEngine(ctx).In("id", userIDs).Find(&usersMap); err != nil {
+		return nil, fmt.Errorf("getCollaborations: %w", err)
```
Comment (denyskon) [inline models/repo/collaboration.go]: Wrong error message?

**Review by denyskon — APPROVED** [review]

File: tests/integration/mirror_pull_test.go (hunk 12 above)
Comment (lunny) [reply tests/integration/mirror_pull_test.go:66]: `findOptions` will be reused on following code.

**Pull request merged**